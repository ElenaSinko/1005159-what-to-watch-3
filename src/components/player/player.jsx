import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {history} from "../../utils";
import {getFilmCards} from "../../reducer/application-state/selectors";

const FULL_SCREEN_SIZE = 100 + `%`;

class Player extends PureComponent {
  constructor(props) {
    super(props);
    this._videoRef = React.createRef();

    this._wholeDuration = null;


    this.state = {
      isPlaying: true,
      progress: null,
      duration: 0.1,
    };
    this._requestFullscreen = this._requestFullscreen.bind(this);
  }

  _handlePlayButtonClick() {
    this.setState((prevState) => ({
      isPlaying: !prevState.isPlaying,
    }));
  }

  _getVideoDuration(wholeDuration) {
    this._wholeDuration = wholeDuration;
  }

  _getCurrentDuration(currentDuration) {
    this.setState({
      progress: currentDuration,
    });
  }

  _requestFullscreen() {
    const videoParent = this._videoRef.current.parentElement;

    if (document.fullscreenElement) {
      document.exitFullscreen();
    } else {
      videoParent.requestFullscreen({navigationUI: `show`});
    }
  }


  componentDidMount() {
    const {isPlaying} = this.state;
    const video = this._videoRef.current;

    video.oncanplaythrough = () => {
      this.setState({isLoading: false});
      this._getVideoDuration(Math.floor(video.duration));
    };
    video.ontimeupdate = () => {
      this.setState({
        progress: Math.floor(video.currentTime),
      });
      this._getCurrentDuration(Math.floor(video.currentTime));
    };
    return isPlaying ? video.play() : video.pause();
  }

  render() {
    const {isPlaying} = this.state;
    const {filmCards, id} = this.props;
    const filmToPlay = filmCards.filter((it) => it.id === parseInt(id, 10))[0];
    const lessDuration = this._wholeDuration - this.state.progress;
    const hours = Math.floor(lessDuration / 3600);
    const minutes = Math.floor((lessDuration - hours * 3600) / 60);
    const seconds = lessDuration % 60;
    const progressBar = this.state.progress * 100 / this._wholeDuration;
    return (
      <div className="player">
        <video
          ref={this._videoRef}
          className="player__video"
          poster={filmToPlay.poster}
          width={FULL_SCREEN_SIZE}
          height={FULL_SCREEN_SIZE}
          src={filmToPlay.src}
          autoPlay={true}
        />
        <button
          onClick={() => {
            history.goBack();
          }}
          type="button"
          className="player__exit"
        >
          Exit
        </button>
        <div className="player__controls">
          <div className="player__controls-row">
            <div className="player__time">
              <progress
                className="player__progress"
                value={progressBar}
                max="100"
              ></progress>
              <div className="player__toggler" style={{left: `${progressBar}` + `%`}}>
                Toggler
              </div>
            </div>
            <div className="player__time-value">{hours}:{minutes}:{seconds}</div>
          </div>
          <div className="player__controls-row">
            <button
              onClick={() => this._handlePlayButtonClick()}
              type="button" className="player__play">
              {isPlaying
                ? <React.Fragment>
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#play"></use>
                  </svg>
                  <span>Play</span>
                </React.Fragment>
                : <React.Fragment>
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#pause"></use>
                  </svg>
                  <span>Pause</span>
                </React.Fragment>}
            </button>
            <div className="player__name">{filmToPlay.name}</div>

            <button onClick={this._requestFullscreen} type="button" className="player__full-screen">
              <svg viewBox="0 0 27 27" width="27" height="27">
                <use xlinkHref="#full-screen"></use>
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

Player.propTypes = {
  isPlaying: PropTypes.bool,
  filmCards: PropTypes.array,
  id: PropTypes.number,
};

const mapStateToProps = (state) => ({
  filmCards: getFilmCards(state),
});

const connectedComponent = connect(mapStateToProps)(Player);
export {connectedComponent as Player};

