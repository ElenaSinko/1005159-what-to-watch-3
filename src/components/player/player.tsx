import * as React from "react";
import {connect} from "react-redux";
import {history} from "../../utils";
import {getFilmCards} from "../../reducer/application-state/selectors";
import {FilmCard} from "../../types";

const FULL_SCREEN_SIZE = 100 + `%`;

interface Props {
  filmCards: FilmCard[];
  isPlaying: boolean;
  id: number | string;
}

interface State {
  isPlaying: boolean;
  isLoading: boolean;
  progress: null | number;
  duration: number;
}

class Player extends React.PureComponent<Props, State> {
  private videoRef: React.RefObject<HTMLVideoElement>;
  private wholeDuration: number;
  constructor(props) {
    super(props);
    this.videoRef = React.createRef();

    this.wholeDuration = null;


    this.state = {
      isPlaying: true,
      isLoading: true,
      progress: null,
      duration: 0.1,
    };
    this._requestFullscreen = this._requestFullscreen.bind(this);
  }

  _handlePlayButtonClick() {
    this.setState((prevState) => ({
      isPlaying: !prevState.isPlaying,
    }));
  }

  _getVideoDuration(wholeDuration) {
    this.wholeDuration = wholeDuration;
  }

  _getCurrentDuration(currentDuration) {
    this.setState({
      progress: currentDuration,
    });
  }

  _requestFullscreen() {
    const videoParent = this.videoRef.current.parentElement;

    if (document.fullscreenElement) {
      document.exitFullscreen();
    } else {
      videoParent.requestFullscreen({navigationUI: `show`});
    }
  }


  componentDidMount() {
    const {isPlaying} = this.state;
    const video = this.videoRef.current;

    if (video) {
      video.oncanplaythrough = () => {
        this.setState({isLoading: false});
        this._getVideoDuration(Math.floor(video.duration));
      };
    }
    if (video) {
      video.ontimeupdate = () => {
        this.setState({
          progress: Math.floor(video.currentTime),
        });
        this._getCurrentDuration(Math.floor(video.currentTime));
      };
      return isPlaying ? video.play() : video.pause();
    }
    return true;
  }

  render() {
    const {isPlaying} = this.state;
    const {filmCards, id} = this.props;
    const filmToPlay = filmCards.filter((it) => it.id === parseInt(`${id}`, 10))[0];
    const lessDuration = this.wholeDuration - this.state.progress;
    const hours = Math.floor(lessDuration / 3600);
    const minutes = Math.floor((lessDuration - hours * 3600) / 60);
    const seconds = lessDuration % 60;
    const progressBar = this.state.progress * 100 / this.wholeDuration;
    return (
      <div className="player">
        <video
          ref={this.videoRef}
          className="player__video"
          poster={filmToPlay.imgPrev}
          width={FULL_SCREEN_SIZE}
          height={FULL_SCREEN_SIZE}
          src={filmToPlay.src}
          autoPlay={true}
        />
        <button
          onClick={() => {
            history.goBack();
          }}
          type="button"
          className="player__exit"
        >
          Exit
        </button>
        <div className="player__controls">
          <div className="player__controls-row">
            <div className="player__time">
              <progress
                className="player__progress"
                value={progressBar}
                max="100"
              ></progress>
              <div className="player__toggler" style={{left: `${progressBar}` + `%`}}>
                Toggler
              </div>
            </div>
            <div className="player__time-value">{hours}:{minutes}:{seconds}</div>
          </div>
          <div className="player__controls-row">
            <button
              onClick={() => this._handlePlayButtonClick()}
              type="button" className="player__play">
              {isPlaying
                ? <React.Fragment>
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#play"></use>
                  </svg>
                  <span>Play</span>
                </React.Fragment>
                : <React.Fragment>
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#pause"></use>
                  </svg>
                  <span>Pause</span>
                </React.Fragment>}
            </button>
            <div className="player__name">{filmToPlay.name}</div>

            <button onClick={this._requestFullscreen} type="button" className="player__full-screen">
              <svg viewBox="0 0 27 27" width="27" height="27">
                <use xlinkHref="#full-screen"></use>
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filmCards: getFilmCards(state),
});

const connectedComponent = connect(mapStateToProps)(Player);
export {connectedComponent as Player};

