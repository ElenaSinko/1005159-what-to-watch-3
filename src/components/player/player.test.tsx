import * as React from "react";
import * as renderer from "react-test-renderer";
import {Player} from "./player";
import {Provider} from "react-redux";
import NameSpace from "../../reducer/name-space";

import configureStore from "redux-mock-store";

const mockStore = configureStore([]);

const cards = [
  {
    name: `Gangs of new york`,
    img: `https://htmlacademy-react-3.appspot.com/wtw/static/film/poster/Gangs_of_New_York_Poster.jpg`,
    imgPrev: `https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg`,
    movieBG: `https://htmlacademy-react-3.appspot.com/wtw/static/film/background/gangs_of_new_york.jpg`,
    BGColor: `#A6B7AC`,
    description: `In 1862, Amsterdam Vallon returns to the Five Points area of New York City seeking revenge against Bill the Butcher, his father's killer.`,
    rating: 8.8,
    movieRatingCount: 370881,
    director: `Martin Scorsese`,
    starring: [`Leonardo DiCaprio`, `Cameron Diaz`, `Daniel Day-Lewis`],
    duration: 167,
    genre: `Crime`,
    released: 2002,
    id: 1,
    movieYear: 2000,
    isFavorite: false,
    srcFullVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  }, {
    name: `Gangs of new york`,
    img: `https://htmlacademy-react-3.appspot.com/wtw/static/film/poster/Gangs_of_New_York_Poster.jpg`,
    imgPrev: `https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg`,
    movieBG: `https://htmlacademy-react-3.appspot.com/wtw/static/film/background/gangs_of_new_york.jpg`,
    BGColor: `#A6B7AC`,
    description: `In 1862, Amsterdam Vallon returns to the Five Points area of New York City seeking revenge against Bill the Butcher, his father's killer.`,
    rating: 8.8,
    movieRatingCount: 370881,
    director: `Martin Scorsese`,
    starring: [`Leonardo DiCaprio`, `Cameron Diaz`, `Daniel Day-Lewis`],
    duration: 167,
    genre: `Crime`,
    released: 2002,
    id: 2,
    movieYear: 2000,
    isFavorite: false,
    srcFullVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  }
];


it(`Render Player`, () => {
  const store = mockStore({
    [NameSpace.APPLICATION_STATE]: {
      filmCards: cards,
    },
  });
  const tree = renderer
    .create(<Provider store={store}><Player
      filmCards={cards}
      id={1}
    /></Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
