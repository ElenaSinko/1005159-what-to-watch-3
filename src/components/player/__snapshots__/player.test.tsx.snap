// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Player 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    height="100%"
    poster="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg"
    src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
    width="100%"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={NaN}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "NaN%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        0
        :
        0
        :
        0
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="21"
          viewBox="0 0 14 21"
          width="14"
        >
          <use
            xlinkHref="#play"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Gangs of new york
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
